File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/picker_dialog_datetime/DateTimePickerDialog.java
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_datetime;

import android.content.Context;
import android.content.res.Configuration;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TimePicker;

import com.github.rooneyandshadows.java.commons.date.DateUtils;
import com.github.rooneyandshadows.java.commons.string.StringUtils;
import com.prolificinteractive.materialcalendarview.CalendarDay;
import com.prolificinteractive.materialcalendarview.MaterialCalendarView;
import com.github.rooneyandshadows.lightbulb.commons.utils.ResourceUtils;
import com.github.rooneyandshadows.lightbulb.dialogs.R;
import com.github.rooneyandshadows.lightbulb.dialogs.base.BasePickerDialogFragment;

import java.util.Date;
import java.util.Locale;

import androidx.appcompat.widget.AppCompatImageButton;
import androidx.appcompat.widget.AppCompatTextView;


public class DateTimePickerDialog extends BasePickerDialogFragment<Date> {
    private static final String SHOWING_TIME_PICKER_TAG = "SHOWING_TIME_PICKER_TAG";
    private static final String DATE_FORMAT_TAG = "DATE_FORMAT_TEXT";
    private static final String DATE_SELECTION_TAG = "DATE_PICKER_SELECTION_TAG";
    private static final String DATE_SELECTION_DRAFT_TAG = "DATE_PICKER_SELECTION_DRAFT_TAG";
    private String dateFormat = "dd MMM HH:mm, yyyy";
    private boolean showingTimePicker;
    private MaterialCalendarView calendarView;
    private TimePicker timePickerView;
    private AppCompatTextView selectionTextValue;
    private AppCompatImageButton modeChangeButton;

    public static DateTimePickerDialog newInstance(
            DialogButtonConfiguration positive, DialogButtonConfiguration negative,
            String dateFormat, boolean cancelable, DialogAnimationTypes animationType) {
        DateTimePickerDialog dialogFragment = new DateTimePickerDialog();
        DialogBundleHelper bundleHelper = new DialogBundleHelper()
                .withPositiveButtonConfig(positive)
                .withNegativeButtonConfig(negative)
                .withCancelable(cancelable)
                .withShowing(false)
                .withDialogType(DialogTypes.NORMAL)
                .withAnimation(animationType == null ? DialogAnimationTypes.NO_ANIMATION : animationType);
        bundleHelper.getBundle().putString(DATE_FORMAT_TAG, dateFormat);
        dialogFragment.setArguments(bundleHelper.getBundle());
        return dialogFragment;
    }

    public DateTimePickerDialog() {
        super(new DateTimeSelection(null, null));
    }

    @Override
    protected void create(Bundle dialogArguments, Bundle savedInstanceState) {
        if (savedInstanceState == null) {
            if (dialogArguments == null) {
                throw new IllegalArgumentException("Bundle args required");
            }
            dateFormat = StringUtils.getOrDefault(dialogArguments.getString(DATE_FORMAT_TAG), dateFormat);
            showingTimePicker = false;
            if (hasSelection()) {
                selection.setCurrentSelection(selection.getCurrentSelection());
            } else {
                Date selectedDateFromArguments = DateUtils.getDateFromStringInDefaultFormat(dialogArguments.getString(DATE_SELECTION_TAG));
                selection.setCurrentSelection(selectedDateFromArguments);
            }
        } else {
            showingTimePicker = savedInstanceState.getBoolean(SHOWING_TIME_PICKER_TAG);
            selection.setCurrentSelection(DateUtils.getDateFromStringInDefaultFormat(savedInstanceState.getString(DATE_SELECTION_TAG)), false);
            selection.setDraftSelection(DateUtils.getDateFromStringInDefaultFormat(savedInstanceState.getString(DATE_SELECTION_DRAFT_TAG)), false);
        }
    }

    @Override
    protected void saveInstanceState(Bundle outState) {
        super.saveInstanceState(outState);
        if (selection.getCurrentSelection() != null)
            outState.putString(DATE_SELECTION_TAG, DateUtils.getDateStringInDefaultFormat(selection.getCurrentSelection()));
        if (selection.getDraftSelection() != null)
            outState.putString(DATE_SELECTION_DRAFT_TAG, DateUtils.getDateStringInDefaultFormat(selection.getDraftSelection()));
        outState.putBoolean(SHOWING_TIME_PICKER_TAG, showingTimePicker);
    }

    @Override
    protected View setDialogLayout(LayoutInflater inflater) {
        int orientation = getResources().getConfiguration().orientation;
        if (orientation == Configuration.ORIENTATION_PORTRAIT)
            return View.inflate(getContext(), R.layout.dialog_picker_datetime_vertical, null);
        else
            return View.inflate(getContext(), R.layout.dialog_picker_datetime_horizontal, null);
    }

    @Override
    protected void configureContent(View view, Bundle savedInstanceState) {
        selectionTextValue = view.findViewById(R.id.dateSelectionValue);
        timePickerView = view.findViewById(R.id.timePickerView);
        modeChangeButton = view.findViewById(R.id.datePickerModeChangeButton);
        calendarView = view.findViewById(R.id.dateCalendarView);
        Context ctx = modeChangeButton.getContext();
        modeChangeButton.setBackgroundDrawable(ResourceUtils.getDrawable(ctx, R.drawable.background_round_corners_transparent));
        modeChangeButton.setOnClickListener(v -> {
            showingTimePicker = !showingTimePicker;
            syncPickerMode();
        });
        modeChangeButton.setVisibility((selection.getDraftSelection() != null) ? View.VISIBLE : View.GONE);
        calendarView.getLeftArrow().setTint(ResourceUtils.getColorByAttribute(ctx, R.attr.colorAccent));
        calendarView.getRightArrow().setTint(ResourceUtils.getColorByAttribute(ctx, R.attr.colorAccent));
        calendarView.setOnDateChangedListener((widget, date, selected) -> {
            if (isDialogShown()) {
                Date draftDate = selection.getDraftSelection();
                int hour = 0;
                int minute = 0;
                int second = 0;
                if (draftDate != null) {
                    hour = DateUtils.getHourOfDay(draftDate);
                    minute = DateUtils.getMinuteOfHour(draftDate);
                }
                selection.setDraftSelection(DateUtils.date(date.getYear(), date.getMonth(), date.getDay(), hour, minute, second));
            } else {
                Date currentDate = selection.getCurrentSelection();
                int hour = 0;
                int minute = 0;
                int second = 0;
                if (currentDate != null) {
                    hour = DateUtils.getHourOfDay(currentDate);
                    minute = DateUtils.getMinuteOfHour(currentDate);
                }
                selection.setCurrentSelection(DateUtils.date(date.getYear(), date.getMonth(), date.getDay(), hour, minute, second));
            }
        });
        timePickerView.setOnTimeChangedListener((view1, hourOfDay, minute) -> {
            Date date = DateUtils.setTimeToDate(selection.getDraftSelection(), hourOfDay, minute, 0);
            if (isDialogShown()) selection.setDraftSelection(date);
            else selection.setCurrentSelection(date);
        });
        syncPickerMode();
        synchronizeSelectUi();
    }

    @Override
    protected void synchronizeSelectUi() {
        Date newDate = selection.hasDraftSelection() ? selection.getDraftSelection() : selection.getCurrentSelection();
        if (newDate != null) {
            if (calendarView != null) {
                modeChangeButton.setVisibility(View.VISIBLE);
                CalendarDay selectedDate = dateToCalendarDay(newDate);
                calendarView.clearSelection();
                calendarView.setDateSelected(selectedDate, true);
                calendarView.setCurrentDate(selectedDate, false);
                timePickerView.setHour(DateUtils.getHourOfDay(newDate));
                timePickerView.setMinute(DateUtils.getMinuteOfHour(newDate));
            }
        } else {
            if (calendarView != null) {
                modeChangeButton.setVisibility(View.GONE);
                calendarView.clearSelection();
            }
        }
        updateHeader(newDate);
    }

    private void updateHeader(Date newDate) {
        if (selectionTextValue != null) {
            Context ctx = selectionTextValue.getContext();
            String dateString = DateUtils.getDateString(dateFormat, newDate, Locale.getDefault());
            if (dateString == null || dateString.equals(""))
                dateString = ResourceUtils.getPhrase(ctx, R.string.dialog_month_picker_empty_text);
            selectionTextValue.setText(dateString);
        }
    }

    private CalendarDay dateToCalendarDay(Date date) {
        if (date == null)
            return null;
        int year = DateUtils.extractYearFromDate(date);
        int month = DateUtils.extractMonthOfYearFromDate(date);
        int day = DateUtils.extractDayOfMonthFromDate(date);
        return CalendarDay.from(year, month, day);
    }

    private void syncPickerMode() {
        Drawable modeIcon = ResourceUtils.getDrawable(modeChangeButton.getContext(), showingTimePicker ? R.drawable.calendar_icon : R.drawable.time_icon);
        modeIcon.setTint(ResourceUtils.getColorByAttribute(modeChangeButton.getContext(), R.attr.colorOnPrimary));
        modeChangeButton.setImageDrawable(modeIcon);
        if (showingTimePicker) {
            timePickerView.setVisibility(View.VISIBLE);
            calendarView.setVisibility(View.GONE);
        } else {
            timePickerView.setVisibility(View.GONE);
            calendarView.setVisibility(View.VISIBLE);
        }
        measureDialogLayout();
    }
}

File: lightbulb-dialogs/src/main/java/com/github/rooneyandshadows/lightbulb/dialogs/picker_dialog_datetime/DateTimePickerDialog.kt
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_datetime

import android.content.res.Configuration
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.TimePicker
import androidx.appcompat.widget.AppCompatImageButton
import androidx.appcompat.widget.AppCompatTextView
import com.github.rooneyandshadows.java.commons.date.DateUtilsOffsetDate
import com.github.rooneyandshadows.lightbulb.commons.utils.ResourceUtils
import com.github.rooneyandshadows.lightbulb.dialogs.R
import com.github.rooneyandshadows.lightbulb.dialogs.base.BasePickerDialogFragment
import com.github.rooneyandshadows.lightbulb.dialogs.base.constraints.regular.RegularDialogConstraints
import com.github.rooneyandshadows.lightbulb.dialogs.base.constraints.regular.RegularDialogConstraintsBuilder
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.DialogTypes
import com.prolificinteractive.materialcalendarview.CalendarDay
import com.prolificinteractive.materialcalendarview.MaterialCalendarView
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.util.*

@Suppress("UNUSED_PARAMETER")
class DateTimePickerDialog : BasePickerDialogFragment<OffsetDateTime>(DateTimeSelection(null, null)) {
    private var showingTimePicker = false
    private var calendarView: MaterialCalendarView? = null
    private var timePickerView: TimePicker? = null
    private var selectionTextValue: AppCompatTextView? = null
    private var modeChangeButton: AppCompatImageButton? = null
    private var zoneOffset = DEFAULT_OFFSET
    var dialogDateFormat = DEFAULT_DATE_FORMAT
        private set
    override var dialogType: DialogTypes
        get() = DialogTypes.NORMAL
        set(value) {}

    companion object {
        private val DEFAULT_OFFSET = ZoneOffset.of(DateUtilsOffsetDate.getLocalTimeZone())
        private const val DEFAULT_DATE_FORMAT = "dd MMM HH:mm, yyyy"
        private const val SHOWING_TIME_PICKER_TAG = "SHOWING_TIME_PICKER_TAG"
        private const val DATE_FORMAT_TAG = "DATE_FORMAT_TEXT"
        private const val DATE_OFFSET_TAG = "DATE_OFFSET_TAG"
        fun newInstance(): DateTimePickerDialog {
            return DateTimePickerDialog()
        }
    }

    @Override
    override fun getRegularConstraints(): RegularDialogConstraints {
        return RegularDialogConstraintsBuilder(this)
            .default()
            .withMaxHeightInPercents(90)
            .build()
    }

    @Override
    override fun doOnSaveDialogProperties(outState: Bundle) {
        super.doOnSaveDialogProperties(outState)
        outState.apply {
            putString(DATE_OFFSET_TAG, zoneOffset.toString())
            putBoolean(SHOWING_TIME_PICKER_TAG, showingTimePicker)
            putString(DATE_FORMAT_TAG, dialogDateFormat)
        }
    }

    @Override
    override fun doOnRestoreDialogProperties(savedState: Bundle) {
        super.doOnRestoreDialogProperties(savedState)
        savedState.apply {
            showingTimePicker = getBoolean(SHOWING_TIME_PICKER_TAG)
            dialogDateFormat = getString(DATE_FORMAT_TAG, dialogDateFormat)
            zoneOffset = ZoneOffset.of(getString(DATE_OFFSET_TAG))
        }
    }

    @Override
    override fun getDialogLayout(layoutInflater: LayoutInflater): View {
        val orientation = resources.configuration.orientation
        return if (orientation == Configuration.ORIENTATION_PORTRAIT)
            layoutInflater.inflate(R.layout.dialog_picker_datetime_vertical, null)
        else layoutInflater.inflate(R.layout.dialog_picker_datetime_horizontal, null)
    }

    @Override
    override fun setupDialogContent(view: View, savedInstanceState: Bundle?) {
        val activeSelection = dialogSelection.getActiveSelection()
        selectViews(view)
        setupHeader(activeSelection)
        val context = requireContext()
        modeChangeButton?.apply {
            background = ResourceUtils.getDrawable(context, R.drawable.dialogs_bg_round_corners_transparent_ripple)
            setOnClickListener {
                showingTimePicker = !showingTimePicker
                syncPickerMode()
            }
        }
        calendarView?.apply {
            leftArrow.setTint(ResourceUtils.getColorByAttribute(context, R.attr.colorAccent))
            rightArrow.setTint(ResourceUtils.getColorByAttribute(context, R.attr.colorAccent))
            if (activeSelection == null) clearSelection()
            else {
                val selectedDate = dateToCalendarDay(activeSelection)
                setDateSelected(selectedDate, true)
                setCurrentDate(selectedDate, false)
            }
        }
        activeSelection?.apply selectedDate@{
            timePickerView?.apply {
                hour = DateUtilsOffsetDate.getHourOfDay(this@selectedDate)
                minute = DateUtilsOffsetDate.getMinuteOfHour(this@selectedDate)
            }
        }
        syncPickerMode()
    }

    private fun checkIfCalendarNeedsSync(currentValue: OffsetDateTime?): Boolean {
        val calendarValue = calendarView!!.selectedDate?.let {
            calendarDayToDate(it, currentValue ?: DateUtilsOffsetDate.nowLocal())
        }
        if (currentValue == null && calendarValue == null) return false
        if (currentValue == null || calendarValue == null) return true
        return !DateUtilsOffsetDate.isDateEqual(currentValue, calendarValue)
    }

    @Override
    override fun onSelectionChange(newSelection: OffsetDateTime?) {
        setupHeader(newSelection)
        if (!checkIfCalendarNeedsSync(newSelection)) return
        calendarView?.apply {
            if (newSelection == null) clearSelection()
            else {
                val selectedDate = dateToCalendarDay(newSelection)
                setDateSelected(selectedDate, true)
                setCurrentDate(selectedDate, false)
            }
        }
        timePickerView?.apply {
            if (newSelection == null) return@apply
            hour = DateUtilsOffsetDate.getHourOfDay(newSelection)
            minute = DateUtilsOffsetDate.getMinuteOfHour(newSelection)
        }
    }

    @Override
    override fun doOnViewStateRestored(savedInstanceState: Bundle?) {
        super.doOnViewStateRestored(savedInstanceState)
        calendarView!!.setOnDateChangedListener { _: MaterialCalendarView?, date: CalendarDay, _: Boolean ->
            if (isDialogShown) {
                val draftDate = dialogSelection.getDraftSelection()
                val hour: Int
                val minute: Int
                val second = 0
                if (draftDate != null) {
                    hour = DateUtilsOffsetDate.getHourOfDay(draftDate)
                    minute = DateUtilsOffsetDate.getMinuteOfHour(draftDate)
                } else {
                    val now = DateUtilsOffsetDate.nowLocal()
                    hour = DateUtilsOffsetDate.getHourOfDay(now)
                    minute = DateUtilsOffsetDate.getMinuteOfHour(now)
                }
                dialogSelection.setDraftSelection(
                    DateUtilsOffsetDate.date(date.year, date.month, date.day, hour, minute, second, zoneOffset)
                )
            } else {
                val currentDate = dialogSelection.getCurrentSelection()
                val hour: Int
                val minute: Int
                val second = 0
                if (currentDate != null) {
                    hour = DateUtilsOffsetDate.getHourOfDay(currentDate)
                    minute = DateUtilsOffsetDate.getMinuteOfHour(currentDate)
                } else {
                    val now = DateUtilsOffsetDate.nowLocal()
                    hour = DateUtilsOffsetDate.getHourOfDay(now)
                    minute = DateUtilsOffsetDate.getMinuteOfHour(now)
                }
                dialogSelection.setCurrentSelection(
                    DateUtilsOffsetDate.date(date.year, date.month, date.day, hour, minute, second, zoneOffset)
                )
            }
        }
        timePickerView!!.setOnTimeChangedListener { _: TimePicker?, hourOfDay: Int, minute: Int ->
            val date = DateUtilsOffsetDate.setTimeToDate(dialogSelection.getDraftSelection(), hourOfDay, minute, 0)
            if (isDialogShown) dialogSelection.setDraftSelection(date) else dialogSelection.setCurrentSelection(date)
        }
    }

    @Override
    override fun setSelection(newSelection: OffsetDateTime?) {
        super.setSelection(newSelection)
        zoneOffset = newSelection?.offset ?: DEFAULT_OFFSET
    }

    fun setDialogDateFormat(dateFormat: String?) {
        val activeSelection = dialogSelection.getActiveSelection()
        dialogDateFormat = dateFormat ?: DEFAULT_DATE_FORMAT
        setupHeader(activeSelection)
    }

    private fun selectViews(view: View) {
        selectionTextValue = view.findViewById(R.id.dateSelectionValue)
        timePickerView = view.findViewById(R.id.timePickerView)
        modeChangeButton = view.findViewById(R.id.datePickerModeChangeButton)
        calendarView = view.findViewById(R.id.dateCalendarView)
    }

    private fun setupHeader(selection: OffsetDateTime?) {
        modeChangeButton?.apply {
            visibility = if (selection == null) View.GONE else View.VISIBLE
        }
        selectionTextValue?.apply {
            val dateString = DateUtilsOffsetDate.getDateString(dialogDateFormat, selection, Locale.getDefault())
            val default = ResourceUtils.getPhrase(requireContext(), R.string.dialog_month_picker_empty_text)
            text = if (dateString.isNullOrBlank()) default else dateString
        }
    }

    private fun dateToCalendarDay(date: OffsetDateTime?): CalendarDay? {
        if (date == null) return null
        val year = DateUtilsOffsetDate.extractYearFromDate(date)
        val month = DateUtilsOffsetDate.extractMonthOfYearFromDate(date)
        val day = DateUtilsOffsetDate.extractDayOfMonthFromDate(date)
        return CalendarDay.from(year, month, day)
    }

    private fun calendarDayToDate(
        calendarDay: CalendarDay?,
        currentSelection: OffsetDateTime,
    ): OffsetDateTime? {
        if (calendarDay == null) return null
        val hour = currentSelection.hour
        val minute = currentSelection.minute
        return DateUtilsOffsetDate.date(calendarDay.year, calendarDay.month, calendarDay.day, hour, minute, 0)
    }

    private fun syncPickerMode() {
        val activeSelection = dialogSelection.getActiveSelection()
        if (showingTimePicker && activeSelection == null) showingTimePicker = false
        val modeIconRes = if (showingTimePicker) R.drawable.dialogs_ic_calendar else R.drawable.dialogs_ic_time
        val modeIcon = ResourceUtils.getDrawable(modeChangeButton!!.context, modeIconRes).apply {
            this!!.setTint(ResourceUtils.getColorByAttribute(modeChangeButton!!.context, R.attr.colorOnPrimary))
        }
        modeChangeButton!!.setImageDrawable(modeIcon)
        if (showingTimePicker) {
            timePickerView!!.visibility = View.VISIBLE
            calendarView!!.visibility = View.GONE
        } else {
            timePickerView!!.visibility = View.GONE
            calendarView!!.visibility = View.VISIBLE
        }
        measureDialogLayout()
    }
}